gitnexus MCP fails on every child spawn: corrupted npx cache (ENOTEMPTY)

## Symptom

Every Claude session spawned by `claude remote-control` in `secure-agent-pod` logs:

```
[ERROR] MCP server "gitnexus" Server stderr: npm error code ENOTEMPTY
npm error syscall rename
npm error path /home/claude/.npm/_npx/e46929201c1128dd/node_modules/gitnexus
npm error dest /home/claude/.npm/_npx/e46929201c1128dd/node_modules/.gitnexus-P2oInPLH
npm error errno -39
npm error ENOTEMPTY: directory not empty, rename '...gitnexus' -> '....gitnexus-P2oInPLH'
[DEBUG] MCP server "gitnexus": Connection failed after 2518ms: MCP error -32000: Connection closed
```

`npx` is trying to swap in a new gitnexus version (rename old → `.gitnexus-XXXXXXXX`), and the destination already exists from a previously interrupted swap. Self-perpetuating: each spawn re-attempts and re-fails.

## Impact

- `gitnexus` MCP server is unavailable in every child session in the pod.
- Other MCPs still load (claude.ai Notion/Drive/Vercel/Netlify/brv all OK).
- Session itself survives the failure — not session-killing, just feature-killing.

## Quick remediation (manual, in-pod)

```bash
rm -rf /home/claude/.npm/_npx/e46929201c1128dd/node_modules/{gitnexus,.gitnexus-*}
```

Next child spawn re-extracts cleanly.

## Better fix (in-image)

`/opt/scripts/npm-cache-prune.sh` (which already runs on a schedule) should also sweep stale `.<package>-<random>` siblings in `~/.npm/_npx/*/node_modules/` — these are atomic-swap remnants that npm sometimes leaves behind after a crash/SIGKILL and that block all future npx-install of the same package.

## Surfaced during

2026-05-23 persistent-agent debugging — willikins session was crashing App-attach children. Manual restart resolved the App issue; the gitnexus failure was found in the per-child debug log while diagnosing and is a separate, pre-existing problem.

--- a/kali/scripts/npm-cache-prune.sh
+++ b/kali/scripts/npm-cache-prune.sh
@@ -9,7 +9,10 @@
 #      for npm cache content is when it was downloaded — slightly more
 #      conservative but correct).
 #   2. Removes cache files untouched for >7d.
-#   3. Runs `npm cache verify` so npm rebuilds its index after the deletion and
+#   3. Sweeps stale npx atomic-swap remnants (`.<pkg>-<random>` dirs) in
+#      ~/.npm/_npx/*/node_modules/ that block all future npx installs of the
+#      same package (agent-images#87).
+#   4. Runs `npm cache verify` so npm rebuilds its index after the deletion and
 #      emits a final size report into the log.
 #
 # A flock guard prevents overlap with a manual `kubectl exec` invocation or a
@@ -71,6 +74,26 @@ if [ -n "${PROBE:-}" ] && [ -f "$PROBE" ]; then
 fi
 
 find "$CACHE_DIR" -type f "$TIME_FLAG" +7 -delete 2>/dev/null || true
+
+# Sweep stale npx atomic-swap remnants (agent-images#87). During an npx
+# package swap, npm renames <pkg> -> .<pkg>-<random>; a crash/SIGKILL mid-swap
+# leaves that dot-dir behind, and every later swap of the same package then
+# fails the rename with ENOTEMPTY — permanently breaking npx install of that
+# package (seen as the gitnexus MCP failing on every child spawn). The regex
+# only matches dash-bearing dot-dirs that are direct children of node_modules
+# (or of an @scope dir), so `.bin` and package-internal dot-dirs never match.
+# The >24h mtime guard ensures an in-flight swap is never raced; mtime (not
+# atime) is deliberate — remnant dirs are never legitimately touched again.
+NPX_DIR="$HOME/.npm/_npx"
+if [ -d "$NPX_DIR" ]; then
+    find "$NPX_DIR" -mindepth 3 -maxdepth 4 -type d -mtime +0 \
+        -regextype posix-extended \
+        -regex '.*/node_modules/(@[^/]+/)?\.[^/]+-[^/]+' -prune -print 2>/dev/null \
+    | while IFS= read -r remnant; do
+        echo "removing stale npx swap remnant: $remnant"
+        rm -rf "$remnant"
+    done
+fi
 
 verify_rc=0
 if command -v npm >/dev/null 2>&1; then

--- a/kali/tests/test_npm_cache_prune.sh
+++ b/kali/tests/test_npm_cache_prune.sh
@@ -62,7 +62,27 @@ bash "$SCRIPT" >"$TMP/out3.log" 2>&1 || rc=$?
 [[ $rc -eq 7 ]] || { echo "FAIL: expected exit 7 from failing npm, got $rc" >&2; cat "$TMP/out3.log" >&2; exit 1; }
 grep -q "WARN: npm cache verify exited 7" "$TMP/out3.log" || { echo "FAIL: warn not logged" >&2; cat "$TMP/out3.log" >&2; exit 1; }
 
-# --- Case 4: lock guard — second concurrent run must noop ---
+# --- Case 4: stale npx swap remnants swept; fresh remnants, real packages,
+# .bin, and package-internal dot-dirs kept ---
+setup_fixture
+run_with_npm_stub 0
+NPX="$AGENT_HOME/.npm/_npx/e46929201c1128dd/node_modules"
+mkdir -p "$NPX/gitnexus" "$NPX/.gitnexus-P2oInPLH" "$NPX/.bin" \
+         "$NPX/@scope/.pkg-AbCdEf12" "$NPX/.fresh-XYZ" "$NPX/gitnexus/.inner-dot"
+echo x > "$NPX/.gitnexus-P2oInPLH/leftover"
+stale_ts="$(date -u -d '2 days ago' +%Y%m%d%H%M)"
+touch -m -t "$stale_ts" "$NPX/.gitnexus-P2oInPLH" "$NPX/@scope/.pkg-AbCdEf12" \
+                        "$NPX/gitnexus/.inner-dot"
+bash "$SCRIPT" >"$TMP/out5.log" 2>&1
+[[ ! -e "$NPX/.gitnexus-P2oInPLH" ]] || { echo "FAIL: stale remnant survived" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+[[ ! -e "$NPX/@scope/.pkg-AbCdEf12" ]] || { echo "FAIL: stale scoped remnant survived" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+[[ -d "$NPX/gitnexus" ]] || { echo "FAIL: real package deleted" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+[[ -d "$NPX/.bin" ]] || { echo "FAIL: .bin deleted" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+[[ -d "$NPX/.fresh-XYZ" ]] || { echo "FAIL: fresh (<24h) remnant deleted" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+[[ -d "$NPX/gitnexus/.inner-dot" ]] || { echo "FAIL: package-internal dot-dir deleted" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+grep -q "removing stale npx swap remnant" "$TMP/out5.log" || { echo "FAIL: remnant removal not logged" >&2; cat "$TMP/out5.log" >&2; exit 1; }
+
+# --- Case 5: lock guard — second concurrent run must noop ---
 setup_fixture
 run_with_npm_stub 0
 # Hold the lock from a background subshell; second invocation should bail.

